File: ios/TasteBuddy/Domain/Models/PersonalTasteInsights.swift
import Foundation

struct PersonalTasteFitDistribution: Codable, Equatable, Sendable {
    var belowPreferred: Int = 0
    var justRight: Int = 0
    var abovePreferred: Int = 0
    var mixed: Int = 0
    var mealCount: Int
}

struct PersonalTasteFitPattern: Codable, Equatable, Identifiable, Sendable {
    let id: String; let attribute: String; let reference: String?; let label: String
    let conditions: [PersonalTasteCondition]; let distribution: PersonalTasteFitDistribution
    let status: String; let repeatedValue: String?
    let mealIDs: [String]; let evidenceIDs: [String]; let evidence: [PersonalTasteEvidence]
}

struct PersonalTasteOverallCell: Codable, Equatable, Sendable {
    let overallValue: String; let attributeValue: String
    var mealIDs: [String]; var evidenceIDs: [String]
}

struct PersonalTasteOverallPattern: Codable, Equatable, Identifiable, Sendable {
    let id: String; let attribute: String; let reference: String?; let label: String
    let conditions: [PersonalTasteCondition]; let cells: [PersonalTasteOverallCell]
    let mealIDs: [String]; let mixedMealIDs: [String]
    let evidenceIDs: [String]; let evidence: [PersonalTasteEvidence]; let causalClaim: String?
}

struct PersonalTasteQuestionImpact: Codable, Equatable, Sendable {
    let changedInterpretationCount: Int
    let alternativeCount: Int
    let comparedMealCount: Int
    let basis: String
}

extension PersonalTasteModelBuilder {
    static let fitValues = ["below_preferred", "just_right", "above_preferred"]
    static let likingValues = ["positive", "neutral", "negative"]
    static let overallValues = ["very_positive", "positive", "neutral", "negative", "very_negative"]

    private struct InsightRating {
        let record: PersonalTasteModelRecord
        let intensity: String?
        var overallValue: String? = nil
        var evidenceRows: [PersonalTasteModelRecord]

        func value(_ dimension: String) -> String? {
            switch dimension {
            case "intensity": return intensity
            case "target": return record.target
            case "phase": return record.phase
            default: return nil
            }
        }
        func matches(_ conditions: [PersonalTasteCondition]) -> Bool {
            conditions.allSatisfy { $0.dimension == "dishKind" ? record.dishKindIDs.contains($0.value) : value($0.dimension) == $0.value }
        }
    }

    static func insightAttributeKey(_ row: PersonalTasteModelRecord) -> String {
        json([nullable(row.attribute), nullable(row.reference)])
    }

    // 소유자·시점·중복 ID 검사를 통과한 관찰에서 같은 선택의 강도만 연결한다.
    private static func scopedRatings(_ records: [PersonalTasteModelRecord], kind: String, scale: String, values: [String]) -> [InsightRating] {
        let scopes = Dictionary(grouping: records, by: scope)
        return records.compactMap { row in
            guard row.kind == kind, row.scale == scale, values.contains(row.value.text),
                  let attribute = row.attribute, !attribute.hasSuffix(".unspecified") else { return nil }
            let related = (scopes[scope(row)] ?? []).filter { sameSelection(row, $0) }
            guard !related.contains(where: { $0.kind == "sensory_presence" && $0.value == .flag(false) }) else { return nil }
            let intensities = related.filter { $0.kind == "sensory_intensity" && $0.scale == "expression-strength-v1" }
            let levels = unique(intensities.map { $0.value.text })
            let intensity = levels.count == 1 && ["weak", "medium", "strong"].contains(levels[0]) ? levels[0] : nil
            return .init(record: row, intensity: intensity, evidenceRows: [row] + intensities)
        }
    }

    private static func insightConditionSets(_ rows: [InsightRating]) -> [(String, [PersonalTasteCondition])] {
        var sets: [String: [PersonalTasteCondition]] = ["[]": []]
        var seenConditions = Set<[PersonalTasteCondition]>()
        for row in rows {
            let dimensions = ["target", "phase", "intensity"].compactMap { dimension -> PersonalTasteCondition? in
                guard let value = row.value(dimension), value != "unspecified" else { return nil }
                return .init(dimension: dimension, value: value)
            }
            let unsorted = dimensions + row.record.dishKindIDs.map { PersonalTasteCondition(dimension: "dishKind", value: $0) }
            guard seenConditions.insert(unsorted).inserted else { continue }
            let available = unsorted.sorted { canonical($0) < canonical($1) }
            for i in available.indices {
                sets[canonical([available[i]])] = [available[i]]
                for j in available.indices where j > i && available[i].dimension != available[j].dimension {
                    let pair = [available[i], available[j]]; sets[canonical(pair)] = pair
                }
            }
        }
        return sets.keys.sorted().map { ($0, sets[$0]!) }
    }

    static func buildFitPatterns(_ records: [PersonalTasteModelRecord], userID: String, policy: PersonalTastePolicy, includeEvidence: Bool = true) -> [PersonalTasteFitPattern] {
        let ratings = scopedRatings(records, kind: "preference_fit", scale: "preference-fit-v1", values: fitValues)
        let attributes = Dictionary(grouping: ratings) { insightAttributeKey($0.record) }
        var patterns: [PersonalTasteFitPattern] = []
        for key in attributes.keys.sorted() {
            let rows = attributes[key]!, first = rows[0].record
            for (conditionKey, conditions) in insightConditionSets(rows) {
                let selected = rows.filter { $0.matches(conditions) }
                let groups = Dictionary(grouping: selected) { $0.record.mealId }
                var distribution = PersonalTasteFitDistribution(mealCount: groups.count)
                var mealValues: [String] = []
                for mealID in groups.keys.sorted() {
                    let values = unique(groups[mealID]!.map { $0.record.value.text })
                    let value = values.count == 1 ? values[0] : "mixed"; mealValues.append(value)
                    switch value {
                    case "below_preferred": distribution.belowPreferred += 1
                    case "just_right": distribution.justRight += 1
                    case "above_preferred": distribution.abovePreferred += 1
                    default: distribution.mixed += 1
                    }
                }
                let values = unique(mealValues), minimum = conditions.isEmpty ? policy.minMeals : policy.minConditionMeals
                let repeated = groups.count >= minimum && values.count == 1 && values[0] != "mixed" ? values[0] : nil
                let status = values.contains("mixed") || values.count > 1 ? "mixed_fit" : repeated != nil ? "repeated_fit" : groups.count == 1 ? "observed_once" : "limited_observations"
                let sources = includeEvidence ? evidence(selected.flatMap(\.evidenceRows)) : []
                patterns.append(.init(id: identifier("fit-pattern", [userID, key, conditionKey]), attribute: first.attribute!, reference: first.reference, label: label(first), conditions: conditions, distribution: distribution, status: status, repeatedValue: repeated, mealIDs: groups.keys.sorted(), evidenceIDs: sources.map(\.id), evidence: sources))
            }
        }
        return patterns.sorted { $0.id < $1.id }
    }

    static func fitQuestions(_ records: [PersonalTasteModelRecord], patterns: [PersonalTasteFitPattern]) -> [Question] {
        scopedRatings(records, kind: "preference_fit", scale: "preference-fit-v1", values: fitValues).flatMap { rating in
            let row = rating.record
            let missing: [(String, Bool, Int)] = [("intensity", rating.intensity == nil, 70), ("target", row.target == "unspecified", 60), ("phase", row.phase == "unspecified", 50)]
            let unstable = patterns.contains { $0.attribute == row.attribute && $0.reference == row.reference && $0.status == "mixed_fit" }
            return missing.filter { $0.1 }.map { facet, _, score in
                Question(attribute: row.attribute!, reference: row.reference, label: label(row), facet: facet, reason: unstable ? "unstable_context_comparison" : "\(facet)_not_reported", score: unstable ? 90 : score, evidenceIDs: rating.evidenceRows.map(\.observationId), mealIDs: [row.mealId], target: row.target, phase: row.phase)
            }
        }
    }

    static func buildOverallPatterns(_ records: [PersonalTasteModelRecord], userID: String, includeEvidence: Bool = true) -> [PersonalTasteOverallPattern] {
        let ratings = scopedRatings(records, kind: "attribute_liking", scale: "attribute-three-category-v1", values: likingValues)
        let overall = Dictionary(grouping: records.filter { $0.kind == "overall_liking" && $0.scale == "overall-five-category-v1" && overallValues.contains($0.value.text) }, by: \.experienceId)
        let pairs = ratings.flatMap { rating in
            (overall[rating.record.experienceId] ?? []).filter { $0.mealId == rating.record.mealId }.map { row in
                var pair = rating; pair.overallValue = row.value.text; pair.evidenceRows.append(row); return pair
            }
        }
        let attributes = Dictionary(grouping: pairs) { insightAttributeKey($0.record) }
        var patterns: [PersonalTasteOverallPattern] = []
        for key in attributes.keys.sorted() {
            let rows = attributes[key]!, first = rows[0].record
            for (conditionKey, conditions) in insightConditionSets(rows) {
                let selected = rows.filter { $0.matches(conditions) }
                let groups = Dictionary(grouping: selected) { $0.record.mealId }
                var cells: [String: PersonalTasteOverallCell] = [:], mixed: [String] = []
                for mealID in groups.keys.sorted() {
                    let items = groups[mealID]!, combinations = unique(items.map { json([$0.overallValue!, $0.record.value.text]) })
                    // 같은 식사의 서로 다른 평가 쌍은 특정 쌍의 반복 근거에 더하지 않는다.
                    guard combinations.count == 1 else { mixed.append(mealID); continue }
                    let cellKey = combinations[0], firstItem = items[0]
                    if cells[cellKey] == nil {
                        cells[cellKey] = .init(overallValue: firstItem.overallValue!, attributeValue: firstItem.record.value.text, mealIDs: [], evidenceIDs: [])
                    }
                    cells[cellKey]!.mealIDs.append(mealID)
                    if includeEvidence { cells[cellKey]!.evidenceIDs.append(contentsOf: items.flatMap { $0.evidenceRows.map(\.observationId) }) }
                }
                let sortedCells = cells.values.map { cell in
                    var result = cell; result.evidenceIDs = unique(cell.evidenceIDs); return result
                }.sorted {
                    let left = overallValues.firstIndex(of: $0.overallValue)!, right = overallValues.firstIndex(of: $1.overallValue)!
                    return left != right ? left < right : likingValues.firstIndex(of: $0.attributeValue)! < likingValues.firstIndex(of: $1.attributeValue)!
                }
                let sources = includeEvidence ? evidence(selected.flatMap(\.evidenceRows)) : []
                patterns.append(.init(id: identifier("overall-pattern", [userID, key, conditionKey]), attribute: first.attribute!, reference: first.reference, label: label(first), conditions: conditions, cells: sortedCells, mealIDs: groups.keys.sorted(), mixedMealIDs: mixed, evidenceIDs: sources.map(\.id), evidence: sources, causalClaim: nil))
            }
        }
        return patterns.sorted { $0.id < $1.id }
    }
}

File: scripts/tba-engine/personal-taste-insights.mjs
import { createHash } from 'node:crypto';

export const FIT_VALUES = ['below_preferred', 'just_right', 'above_preferred'];
export const LIKING_VALUES = ['positive', 'neutral', 'negative'];
export const OVERALL_VALUES = ['very_positive', 'positive', 'neutral', 'negative', 'very_negative'];
export const unique = values => [...new Set(values)].sort();
export const canonical = value => JSON.stringify(value, (_, item) => item && typeof item === 'object' && !Array.isArray(item)
  ? Object.fromEntries(Object.entries(item).sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0)) : item);
export const modelIdentifier = (kind, parts) => `${kind}:${createHash('sha256').update(canonical(parts)).digest('hex').slice(0, 20)}`;
export const recordScope = record => canonical([record.experienceId, record.attribute, record.reference ?? null, record.target, record.phase]);
export function sameSelection(left, right) {
  const a = left.selectionEvidence, b = right.selectionEvidence;
  if (!a || !b) return !a && !b;
  return ['selectionID', 'type', 'catalogVersion', 'relatedBubbleID'].every(key => (a[key] ?? null) === (b[key] ?? null));
}
export const attributeIdentity = row => canonical([row.attribute, row.reference ?? null]);
export const modelLabel = row => row.reference ?? row.attributeLabel ?? ({
  'taste.sweet': '단맛', 'taste.sour': '산미', 'taste.salty': '짠맛', 'taste.bitter': '쓴맛',
  'taste.umami': '감칠맛', 'texture.crisp': '바삭함', 'texture.soft': '부드러운 식감', 'aroma.roasted': '구운 향',
}[row.attribute] ?? row.attribute);
export function grouped(rows, key) {
  const groups = new Map();
  for (const row of rows) { const id = key(row); if (!groups.has(id)) groups.set(id, []); groups.get(id).push(row); }
  return [...groups].sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0);
}
export function modelEvidence(rows) {
  return [...new Map(rows.map(row => [row.observationId, {
    id: row.observationId, mealID: row.mealId, phrase: row.phrase ?? '',
    sourceSpans: row.sourceSpans ?? row.sourceAnswerRefs?.flatMap(ref => ref.sourceSpans ?? []) ?? [],
    selectionEvidence: row.selectionEvidence ?? null, conditionSources: row.conditionSources ?? [],
  }])).values()].sort((a, b) => a.id < b.id ? -1 : a.id > b.id ? 1 : 0);
}
export function conditionSets(rows) {
  const sets = new Map([['[]', []]]), seen = new Set();
  for (const row of rows) {
    const available = ['target', 'phase', 'intensity'].flatMap(dimension => row[dimension] && row[dimension] !== 'unspecified'
      ? [{ dimension, value: row[dimension] }] : []).concat(row.dishKindIDs.map(value => ({ dimension: 'dishKind', value })));
    const signature = canonical(available); if (seen.has(signature)) continue; seen.add(signature);
    available.sort((a, b) => canonical(a) < canonical(b) ? -1 : 1);
    for (let i = 0; i < available.length; i++) {
      sets.set(canonical([available[i]]), [available[i]]);
      for (let j = i + 1; j < available.length; j++) {
        if (available[i].dimension !== available[j].dimension) {
          const pair = [available[i], available[j]]; sets.set(canonical(pair), pair);
        }
      }
    }
  }
  return [...sets].sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0);
}
export const matchesConditions = (row, conditions) => conditions.every(c => c.dimension === 'dishKind'
  ? row.dishKindIDs.includes(c.value) : row[c.dimension] === c.value);

// 이 함수에는 소유자·시점·중복 ID 검사를 통과한 현재 유효 관찰만 전달한다.
export function scopedRatings(records, kind, scale, values) {
  const scopes = new Map(grouped(records, recordScope));
  return records.filter(row => row.kind === kind && row.scale === scale && values.includes(row.value)
    && row.attribute && !row.attribute.endsWith('.unspecified')).flatMap(row => {
    const related = (scopes.get(recordScope(row)) ?? []).filter(other => sameSelection(row, other));
    if (related.some(other => other.kind === 'sensory_presence' && other.value === false)) return [];
    const intensities = related.filter(other => other.kind === 'sensory_intensity' && other.scale === 'expression-strength-v1');
    const levels = unique(intensities.map(other => other.value));
    return [{ ...row, intensity: levels.length === 1 && ['weak', 'medium', 'strong'].includes(levels[0]) ? levels[0] : null,
      intensityConflict: levels.length > 1, evidenceRows: [row, ...intensities] }];
  });
}

export function buildFitPatterns(records, userID, policy, { includeEvidence = true } = {}) {
  const ratings = scopedRatings(records, 'preference_fit', 'preference-fit-v1', FIT_VALUES);
  const patterns = [];
  for (const [key, rows] of grouped(ratings, attributeIdentity)) {
    for (const [conditionKey, conditions] of conditionSets(rows)) {
      const selected = rows.filter(row => matchesConditions(row, conditions));
      const meals = grouped(selected, row => row.mealId).map(([mealID, items]) => {
        const values = unique(items.map(row => row.value));
        return { mealID, value: values.length === 1 ? values[0] : 'mixed' };
      });
      const distribution = { belowPreferred: 0, justRight: 0, abovePreferred: 0, mixed: 0, mealCount: meals.length };
      const fields = { below_preferred: 'belowPreferred', just_right: 'justRight', above_preferred: 'abovePreferred', mixed: 'mixed' };
      for (const meal of meals) distribution[fields[meal.value]]++;
      const values = unique(meals.map(meal => meal.value));
      const minimum = conditions.length ? policy.minConditionMeals : policy.minMeals;
      const repeatedValue = meals.length >= minimum && values.length === 1 && values[0] !== 'mixed' ? values[0] : null;
      const status = values.includes('mixed') || values.length > 1 ? 'mixed_fit'
        : repeatedValue ? 'repeated_fit' : meals.length === 1 ? 'observed_once' : 'limited_observations';
      const evidence = includeEvidence ? modelEvidence(selected.flatMap(row => row.evidenceRows)) : [];
      patterns.push({ id: modelIdentifier('fit-pattern', [userID, key, conditionKey]), attribute: rows[0].attribute,
        reference: rows[0].reference ?? null, label: modelLabel(rows[0]), conditions, distribution, status, repeatedValue,
        mealIDs: meals.map(meal => meal.mealID), evidenceIDs: evidence.map(item => item.id), evidence });
    }
  }
  return patterns.sort((a, b) => a.id < b.id ? -1 : 1);
}

export function buildOverallPatterns(records, userID, { includeEvidence = true } = {}) {
  const ratings = scopedRatings(records, 'attribute_liking', 'attribute-three-category-v1', LIKING_VALUES);
  const overallByExperience = new Map(grouped(records.filter(row => row.kind === 'overall_liking'
    && row.scale === 'overall-five-category-v1' && OVERALL_VALUES.includes(row.value)), row => row.experienceId));
  // 연결의 기본 단위는 식사 전체가 아닌 실제로 평가한 디시다. 다른 디시의 전체 평가를 붙이지 않는다.
  const pairs = ratings.flatMap(row => (overallByExperience.get(row.experienceId) ?? [])
    .filter(overall => overall.mealId === row.mealId)
    .map(overall => ({ ...row, overallValue: overall.value, evidenceRows: [...row.evidenceRows, overall] })));
  const patterns = [];
  for (const [key, rows] of grouped(pairs, attributeIdentity)) {
    for (const [conditionKey, conditions] of conditionSets(rows)) {
      const selected = rows.filter(row => matchesConditions(row, conditions));
      const cells = new Map(), mixedMealIDs = [];
      for (const [mealID, items] of grouped(selected, row => row.mealId)) {
        const combinations = unique(items.map(row => canonical([row.overallValue, row.value])));
        // 한 식사에 서로 다른 평가 쌍이 있으면 어느 쌍도 반복 한 표로 승격하지 않는다.
        if (combinations.length !== 1) { mixedMealIDs.push(mealID); continue; }
        const cellKey = combinations[0], first = items[0];
        const cell = cells.get(cellKey) ?? { overallValue: first.overallValue, attributeValue: first.value, mealIDs: [], evidenceIDs: [] };
        cell.mealIDs.push(mealID); if (includeEvidence) cell.evidenceIDs.push(...items.flatMap(row => row.evidenceRows.map(r => r.observationId)));
        cells.set(cellKey, cell);
      }
      const evidence = includeEvidence ? modelEvidence(selected.flatMap(row => row.evidenceRows)) : [];
      patterns.push({ id: modelIdentifier('overall-pattern', [userID, key, conditionKey]), attribute: rows[0].attribute,
        reference: rows[0].reference ?? null, label: modelLabel(rows[0]), conditions,
        cells: [...cells.values()].map(cell => ({ ...cell, evidenceIDs: unique(cell.evidenceIDs) })).sort((a, b) => OVERALL_VALUES.indexOf(a.overallValue) - OVERALL_VALUES.indexOf(b.overallValue)
          || LIKING_VALUES.indexOf(a.attributeValue) - LIKING_VALUES.indexOf(b.attributeValue)),
        mealIDs: unique(selected.map(row => row.mealId)), mixedMealIDs,
        evidenceIDs: evidence.map(item => item.id), evidence, causalClaim: null });
    }
  }
  return patterns.sort((a, b) => a.id < b.id ? -1 : 1);
}
